synapse/storage/databases/main/relations.py

```diff
@@ -814,6 +814,93 @@ def _get_event_relations(
             "get_event_relations", _get_event_relations
         )
 
+    @cached(tree=True)
+    async def get_threads(
+        self,
+        room_id: str,
+        limit: int = 5,
+        from_token: Optional[StreamToken] = None,
+        to_token: Optional[StreamToken] = None,
+    ) -> Tuple[List[str], Optional[StreamToken]]:
+        """Get a list of thread IDs, ordered by topological ordering of their
+        latest reply.
+
+        Args:
+            room_id: The room the event belongs to.
+            limit: Only fetch the most recent `limit` threads.
+            from_token: Fetch rows from the given token, or from the start if None.
+            to_token: Fetch rows up to the given token, or up to the end if None.
+
+        Returns:
+            A tuple of:
+                A list of thread root event IDs.
+
+                The next stream token, if one exists.
+        """
+        pagination_clause = generate_pagination_where_clause(
+            direction="b",
+            column_names=("topological_ordering", "stream_ordering"),
+            from_token=from_token.room_key.as_historical_tuple()
+            if from_token
+            else None,
+            to_token=to_token.room_key.as_historical_tuple() if to_token else None,
+            engine=self.database_engine,
+        )
+
+        if pagination_clause:
+            pagination_clause = "AND " + pagination_clause
+
+        sql = f"""
+            SELECT relates_to_id, MAX(topological_ordering), MAX(stream_ordering)
+            FROM event_relations
+            INNER JOIN events USING (event_id)
+            WHERE
+                room_id = ? AND
+                relation_type = '{RelationTypes.THREAD}'
+                {pagination_clause}
+            GROUP BY relates_to_id
+            ORDER BY MAX(topological_ordering) DESC, MAX(stream_ordering) DESC
+            LIMIT ?
```

Comment:
Does this have a sane looking query plan?

Reply:
I believe I checked this and it was reasonable, will double check though!

Reply:
```sql
EXPLAIN ANALYZE SELECT relates_to_id, MAX(topological_ordering), MAX(stream_ordering)
FROM event_relations
         INNER JOIN events USING (event_id)
WHERE room_id = '!xxx:matrix.org'
  AND relation_type = 'm.thread'
GROUP BY relates_to_id
ORDER BY MAX(topological_ordering) DESC, MAX(stream_ordering) DESC
LIMIT 20;
                                                                                 QUERY PLAN                                 
                                                
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=398932.50..398932.53 rows=11 width=59) (actual time=20.609..22.838 rows=20 loops=1)
   ->  Sort  (cost=398932.50..398932.53 rows=11 width=59) (actual time=20.607..22.834 rows=20 loops=1)
         Sort Key: (max(events.topological_ordering)) DESC, (max(events.stream_ordering)) DESC
         Sort Method: top-N heapsort  Memory: 29kB
         ->  GroupAggregate  (cost=398932.09..398932.31 rows=11 width=59) (actual time=20.389..22.775 rows=136 loops=1)
               Group Key: event_relations.relates_to_id
               ->  Sort  (cost=398932.09..398932.12 rows=11 width=59) (actual time=20.384..22.638 rows=547 loops=1)
                     Sort Key: event_relations.relates_to_id
                     Sort Method: quicksort  Memory: 101kB
                     ->  Gather  (cost=3841.08..398931.90 rows=11 width=59) (actual time=4.102..22.131 rows=547 loops=1)
                           Workers Planned: 2
                           Workers Launched: 2
                           ->  Nested Loop  (cost=2841.08..397930.80 rows=5 width=59) (actual time=1.302..9.227 rows=182 loops=3)
                                 ->  Parallel Bitmap Heap Scan on events  (cost=2840.51..207353.70 rows=43108 width=55) (actual time=0.858..2.522 rows=773 loops=3)
                                       Recheck Cond: (room_id = '!xxx:matrix.org'::text)
                                       Heap Blocks: exact=1424
                                       ->  Bitmap Index Scan on events_room_stream  (cost=0.00..2814.65 rows=103459 width=0) (actual time=1.558..1.558 rows=2320 loops=1)
                                             Index Cond: (room_id = '!xxx:matrix.org'::text)
                                 ->  Index Scan using event_relations_id on event_relations  (cost=0.57..4.42 rows=1 width=86) (actual time=0.008..0.008 rows=0 loops=2320)
                                       Index Cond: (event_id = events.event_id)
                                       Filter: (relation_type = 'm.thread'::text)
                                       Rows Removed by Filter: 0
 Planning time: 0.640 ms
 Execution time: 22.918 ms
(24 rows)

```

Reply:
@erikjohnston Seems that this might be a bit expensive, although empirically is fast. What do you think?

Reply:
Yeah, alas that is postgres being clever and noticing that the room you've given it doesn't have many events in it, and so is looping through *all* events in the room and comparing that against the `event_relations` table. Trying this out on Matrix HQ on jki.re gives:

```
synapse=# EXPLAIN ANALYZE SELECT relates_to_id, MAX(topological_ordering), MAX(stream_ordering)
FROM event_relations
         INNER JOIN events USING (event_id)
WHERE room_id = '!OGEhHVWSdvArJzumhm:matrix.org'
  AND relation_type = 'm.thread'
GROUP BY relates_to_id
ORDER BY MAX(topological_ordering) DESC, MAX(stream_ordering) DESC
LIMIT 20;
                                                                                QUERY PLAN                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=20041.64..20041.69 rows=20 width=58) (actual time=658.571..671.712 rows=20 loops=1)
   ->  Sort  (cost=20041.64..20042.06 rows=167 width=58) (actual time=658.570..671.707 rows=20 loops=1)
         Sort Key: (max(events.topological_ordering)) DESC, (max(events.stream_ordering)) DESC
         Sort Method: top-N heapsort  Memory: 30kB
         ->  Finalize GroupAggregate  (cost=20016.92..20037.20 rows=167 width=58) (actual time=658.104..671.645 rows=92 loops=1)
               Group Key: event_relations.relates_to_id
               ->  Gather Merge  (cost=20016.92..20034.48 rows=140 width=58) (actual time=658.090..671.540 rows=140 loops=1)
                     Workers Planned: 2
                     Workers Launched: 2
                     ->  Partial GroupAggregate  (cost=19016.90..19018.30 rows=70 width=58) (actual time=652.729..652.785 rows=47 loops=3)
                           Group Key: event_relations.relates_to_id
                           ->  Sort  (cost=19016.90..19017.07 rows=70 width=58) (actual time=652.714..652.725 rows=68 loops=3)
                                 Sort Key: event_relations.relates_to_id
                                 Sort Method: quicksort  Memory: 34kB
                                 Worker 0:  Sort Method: quicksort  Memory: 35kB
                                 Worker 1:  Sort Method: quicksort  Memory: 34kB
                                 ->  Nested Loop  (cost=0.56..19014.75 rows=70 width=58) (actual time=250.692..651.439 rows=68 loops=3)
                                       ->  Parallel Seq Scan on event_relations  (cost=0.00..11135.11 rows=1731 width=84) (actual time=150.095..316.856 rows=2476 loops=3)
                                             Filter: (relation_type = 'm.thread'::text)
                                             Rows Removed by Filter: 162474
                                       ->  Index Scan using events_event_id_key on events  (cost=0.56..4.55 rows=1 width=53) (actual time=0.134..0.134 rows=0 loops=7428)
                                             Index Cond: (event_id = event_relations.event_id)
                                             Filter: (room_id = '!OGEhHVWSdvArJzumhm:matrix.org'::text)
                                             Rows Removed by Filter: 1
 Planning Time: 0.436 ms
 Execution Time: 671.832 ms
```

(Note that it has decided to scan the *entire* `event_relations` table). 

On matrix.org:


```
 EXPLAIN ANALYZE SELECT relates_to_id, MAX(topological_ordering), MAX(stream_ordering)
FROM event_relations
         INNER JOIN events USING (event_id)
WHERE room_id = '!OGEhHVWSdvArJzumhm:matrix.org'
  AND relation_type = 'm.thread'
GROUP BY relates_to_id
ORDER BY MAX(topological_ordering) DESC, MAX(stream_ordering) DESC

                                                                                   QUERY PLAN                                                                                   
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=398933.33..398933.36 rows=11 width=59) (actual time=36183.267..36198.204 rows=20 loops=1)
   ->  Sort  (cost=398933.33..398933.36 rows=11 width=59) (actual time=36183.267..36198.201 rows=20 loops=1)
         Sort Key: (max(events.topological_ordering)) DESC, (max(events.stream_ordering)) DESC
         Sort Method: top-N heapsort  Memory: 29kB
         ->  GroupAggregate  (cost=398932.92..398933.14 rows=11 width=59) (actual time=36183.141..36198.123 rows=92 loops=1)
               Group Key: event_relations.relates_to_id
               ->  Sort  (cost=398932.92..398932.95 rows=11 width=59) (actual time=36183.133..36198.074 rows=203 loops=1)
                     Sort Key: event_relations.relates_to_id
                     Sort Method: quicksort  Memory: 53kB
                     ->  Gather  (cost=3841.08..398932.73 rows=11 width=59) (actual time=10066.571..36197.668 rows=203 loops=1)
                           Workers Planned: 2
                           Workers Launched: 2
                           ->  Nested Loop  (cost=2841.08..397931.63 rows=5 width=59) (actual time=10249.827..36169.760 rows=68 loops=3)
                                 ->  Parallel Bitmap Heap Scan on events  (cost=2840.51..207353.70 rows=43108 width=55) (actual time=749.647..16447.442 rows=191989 loops=3)
                                       Recheck Cond: (room_id = '!OGEhHVWSdvArJzumhm:matrix.org'::text)
                                       Rows Removed by Index Recheck: 6028093
                                       Heap Blocks: exact=6206 lossy=174004
                                       ->  Bitmap Index Scan on events_room_stream  (cost=0.00..2814.65 rows=103459 width=0) (actual time=695.062..695.062 rows=575969 loops=1)
                                             Index Cond: (room_id = '!OGEhHVWSdvArJzumhm:matrix.org'::text)
                                 ->  Index Scan using event_relations_id on event_relations  (cost=0.57..4.42 rows=1 width=86) (actual time=0.102..0.102 rows=0 loops=575968)
                                       Index Cond: (event_id = events.event_id)
                                       Filter: (relation_type = 'm.thread'::text)
                                       Rows Removed by Filter: 0
 Planning time: 13.238 ms
 Execution time: 36198.345 ms
```

I'm not sure how to solve this, as you want to be able to scan an index that has the ordering terms in them (otherwise you have to pull out the *full* result set, sort them and *then* take the limit). The closest I've come is:

```
EXPLAIN ANALYZE  SELECT  relates_to_id, topological_ordering, stream_ordering
FROM event_relations INNER JOIN events USING (event_id)
WHERE room_id = '!OGEhHVWSdvArJzumhm:matrix.org'
  AND relation_type = 'm.thread' ORDER BY topological_ordering DESC, stream_ordering DESC
LIMIT 20;
                                                                                 QUERY PLAN                                                                                  
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1001.30..401909.96 rows=11 width=59) (actual time=178.474..1027.784 rows=20 loops=1)
   ->  Gather Merge  (cost=1001.30..401909.96 rows=11 width=59) (actual time=178.474..1027.775 rows=20 loops=1)
         Workers Planned: 2
         Workers Launched: 1
         ->  Nested Loop  (cost=1.27..400908.67 rows=5 width=59) (actual time=93.417..747.919 rows=14 loops=2)
               ->  Parallel Index Scan Backward using events_order_room on events  (cost=0.70..210330.73 rows=43108 width=55) (actual time=0.021..245.899 rows=3146 loops=2)
                     Index Cond: (room_id = '!OGEhHVWSdvArJzumhm:matrix.org'::text)
               ->  Index Scan using event_relations_id on event_relations  (cost=0.57..4.42 rows=1 width=86) (actual time=0.159..0.159 rows=0 loops=6293)
                     Index Cond: (event_id = events.event_id)
                     Filter: (relation_type = 'm.thread'::text)
                     Rows Removed by Filter: 0
 Planning time: 0.270 ms
 Execution time: 1027.825 ms
(13 rows)
```

Which doesn't aggregate the result set and *still* takes over a second on matrix.org. I think to make this faster you'd need to copy the room ID and ordering columns into `event_relations`

Reply:
I'll play a bit with different queries and see if I can do better. If not we might need to have a threads specific table -- that could help, I think?

Reply:
Yup, having a table with all these values in will allow us to have a good index :+1:

Reply:
Another approach might be to avoid the `MAX(...)` call by just fetching the events and then filtering in Python for the max-per-thread. I think this wouldn't really paginate well though.

If we had a threads-specific table, I guess it would be:

* `thread_id`
* `latest_event_id` (and its `topological_ordering` and `stream_ordering`?)

The issue with this is that it won't give a stable sort, e.g.:

1. Consider threads`A`, `B` and `C` which have latest events (respectively) `M`, `N`, `O`. (Alphabetical = topo / stream ordering).
2. A client requests 2 threads and gets `B`, `C`.
3. An event `X` is created in thread `A`.
4. Fetching to the next page of threads returns nothing since the topo/stream ordering of `X` is past `N`, and `O`.

🤔

Reply:
> Yup, having a table with all these values in will allow us to have a good index 👍

Thinking about this more I think you're suggesting having a table with an index on all the fields we need for look-up: `(room_id, thread_id, event_id, topo_ordering, stream_ordering)`? Possibly `sender` too if we want to reasonably handle ignored users?